baselines/flwr_baselines/publications/wav2vec2/tedlium/preprocessing/data_prepare.py

```diff
@@ -0,0 +1,127 @@
+            with urllib.request.urlopen(
+                url,
+                context=ssl._create_unverified_context(),  # pylint: disable=protected-access
+            ) as response, open(filename, "wb") as out_file:
+                total_size = int(response.getheader("Content-Length"))
+                block_size = 1024 * 8
+                count = 0
+                while True:
+                    data = response.read(block_size)
+                    if not data:
+                        break
+                    count += 1
+                    out_file.write(data)
+                    percent = int(count * block_size * 100 / total_size)
+                    print(
+                        f"\rDownloading: {percent}% [{count * block_size}/{total_size}]",
+                        end="",
+                    )
+                print("\nDownload complete.")
+                break
+        except Exception as error:  # pylint: disable=broad-except
+            print(f"\nError occurred during download: {error}")
+            retries -= 1
+            if retries > 0:
+                print(f"Retrying ({retries} retries left)...")
+            else:
+                print("Download failed.")
+                raise error
+
+
+def _extract_file(filename, extract_path):
+    """Extract the contents and show a progress bar."""
+
+    print(f"Extracting {filename}...")
+    with tarfile.open(filename, "r:gz") as tar:
+        members = tar.getmembers()
+        total_files = len(members)
+        current_file = 0
+        for member in members:
+            current_file += 1
+            tar.extract(member, path=extract_path)
+            percent = int(current_file * 100 / total_files)
+            print(f"\rExtracting: {percent}% [{current_file}/{total_files}]", end="")
+        print("\nExtraction complete.")
+
+
+def _delete_file(filename):
+    """Delete the downloaded file."""
+
+    os.remove(filename)
+    print(f"Deleted {filename}.")
+
+
+def _csv_path_audio():
+    """Change the path corespond to your actual path."""
+
+    df_concat_train = None
+    df_concat_dev = None
+    df_concat_test = None
+    for i in range(1983):
+        df_train = pd.read_csv(f"./data/client_{i}/ted_train.csv")
+        df_dev = pd.read_csv(f"./data/client_{i}/ted_dev.csv")
+        df_test = pd.read_csv(f"./data/client_{i}/ted_test.csv")
+        df_train["wav"] = df_train["wav"].str.replace(
+            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
+        )
+        df_dev["wav"] = df_dev["wav"].str.replace(
+            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
+        )
+        df_test["wav"] = df_test["wav"].str.replace(
+            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
+        )
+        df_train.to_csv(f"./data/client_{i}/ted_train.csv", index=False)
+        df_dev.to_csv(f"./data/client_{i}/ted_dev.csv", index=False)
+        df_test.to_csv(f"./data/client_{i}/ted_test.csv", index=False)
```

Comment:
```suggestion
    for i in range(1984):
        data_path = f"./data/client_{i}/"
        if i == 1983:
            data_path = "./data/server/"
        df_train = pd.read_csv(data_path + ted_train.csv" + "ted_train.csv")
        df_dev = pd.read_csv(data_path + ted_train.csv" + "ted_dev.csv")
        df_test = pd.read_csv(data_path + ted_train.csv" + "ted_test.csv")
        df_train["wav"] = df_train["wav"].str.replace(
            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
        )
        df_dev["wav"] = df_dev["wav"].str.replace(
            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
        )
        df_test["wav"] = df_test["wav"].str.replace(
            "path", "./data/audio/TEDLIUM_release-3/legacy/train/sph/"
        )
        df_train.to_csv(data_path + ted_train.csv" + "ted_train.csv", index=False)
        df_dev.to_csv(data_path + ted_train.csv" + "ted_dev.csv", index=False)
        df_test.to_csv(data_path + ted_train.csv" + "ted_test.csv", index=False)
```

Update new data folder structure ( just add the server folder nothing else)

Reply:
Hi @tuanct1997 , i'm looking into this again. Are you sure the `i==1983` logic is correct? I'm asking because in other parts of the code you assume the `cid` for the servers is `19999`. WIth the above code (with small fixes to properly join the paths passed to `pd.read_csv`) I'm getting errors of the type:

```
ValueError: Duplicate id: GretchenCarlsonandDavidBrooks_2017S_920.64_924.26
```
Since it seems there are many duplicates in `data/server`...
Any ideas how to deal with that. Before making the changes you propose above (and changes to `model.py`), i wasn't seeing that problem

Reply:
Hi @jafermarq ,

I double-checked the updated csv folder on my personal repo (where baseline set-up script takes the csv files). I don't see any duplicate_id for each file. It's quite weird to have that error since each client only take 1 train_csv file to train. I have quick tested on my old version (flower <1.0 with 2 clients) and seem to be working. 

About the logic, it's correct, I just re-format the data with new server folder. So it will be easier for people to quickly access to csv files. However, when re-look at the suggestion, the code was too long and confusing. I suggest to rewrite the` csv_path_audio(path)` function to : 

```suggestion
def csv_path_audio(path):
    for subdir, dirs, files in os.walk("./data"):
        for file in files:
            if file.endswith('.csv'):
                df = pd.read_csv(os.path.join(subdir,file))
                df["wav"] = df["wav"].str.replace("path", path)
                df.to_csv(os.path.join(subdir,file), index = False)
```

Could you please re-test it. If the problem still occur , could you also provide me with the cid?